[transformers ] Add @lit/ts-transformers package with idiomatic decorator transform

aomarks opened the pull request:
Adds a `@lit/ts-transformers` package to contain TypeScript transforms.

This PR starts with the "idiomatic" decorator tansform, which transforms all Lit decorators into JavaScript idiomatic style.

The initial motivating use case for this transform is to automatically generate https://lit.dev examples in non-decorator JavaScript, controlled by a JS/TS switch (https://github.com/lit/lit.dev/issues/332).

There may also be some size/performance improvements by using this transform, but we'll want to benchmark before saying that -- plus there's more we can do for this goal, such as defining getters/setters for reactive properties too.

Fixes https://github.com/lit/lit/issues/1886

## Example

### Before

```ts
import {LitElement, html} from 'lit';
import {
  customElement,
  property,
  state,
  query,
  eventOptions,
} from 'lit/decorators.js';

@customElement('simple-greeting')
class SimpleGreeting extends LitElement {
  @property()
  name = 'Somebody';

  @state()
  private counter = 0;

  @query('#myButton')
  button?: HTMLButtonElement;

  render() {
    return html`
      <button id="myButton" @click=${this._onClick}>
        Hello ${this.name} (${this.counter})
      </button>
    `;
  }

  @eventOptions({capture: true})
  _onClick(event: Event) {
    this.counter++;
    console.log('click', event.target);
  }
}
```

### After

```ts
import {LitElement, html} from 'lit';

class MyElement extends LitElement {
  static get properties() {
    return {
      str: {},
      num: {state: true},
    };
  }

  get button() {
    return this.renderRoot?.querySelector('#myButton');
  }

  constructor() {
    super(...arguments);
    this.name = 'Somebody';
    this.counter = 0;
    this._onClick = {
      handleEvent: (event) => {
        console.log('click', event.target);
      },
      capture: true,
    };
  }

  render() {
    return html`
      <button id="myButton" @click=${this._onClick}>
        Hello ${this.name} (${this.counter})
      </button>
    `;
  }
}

customElements.define('simple-greeting', SimpleGreeting);
```

rictic commented on packages/transformers/src/lit-transformer.ts:
```diff
@@ -0,0 +1,263 @@
+  private _visitImportDeclaration(node: ts.ImportDeclaration) {
+    if (
+      ts.isStringLiteral(node.moduleSpecifier) &&
+      node.moduleSpecifier.text === 'lit/decorators.js'
```
Consider also checking for `'lit-element'` since that's probably still common, and `'`lit/decorators'` since that's how auto-imports will tend to get written

aomarks replied:
Done. Actually I think all of these are valid (!) so I'm mostly using prefix tests now:

```
'lit/decorators.js',
'lit/decorators',
'lit/decorators/custom-element.js',
'lit/decorators/custom-element',
'@lit/reactive-element/decorators.js',
'@lit/reactive-element/decorators',
'@lit/reactive-element/decorators/custom-element.js',
'@lit/reactive-element/decorators/custom-element',
'lit-element',
'lit-element/index.js',
'lit-element/index',
'lit-element/decorators.js',
'lit-element/decorators',
```

justinfagnani replied:
The non-file-extension imports are not valid actually. They could be used by toolchains that don't (yet) support package exports, but I don't think we should support that. Those projects will break and have to fix imports when they update tools.

aomarks replied:
> The non-file-extension imports are not valid actually. They could be used by toolchains that don't (yet) support package exports, but I don't think we should support that. Those projects will break and have to fix imports when they update tools.

If we don't support them, the behavior will be that we silently don't transform decorators if they are imported without an extension. Is that your preference? We could also have an explicit check for that style of import, but here seems like the wrong place to do that.

justinfagnani replied:
I feel like we should error here. At first we'll be the ones using this and our code should be correct. If we find that we see valid requests to support extension-less imports, and it somehow would be better to fail, and it guaranteed to fail, somewhere else in the pipeline, then we could relax the strictness here. I don't want to perpetuate tools that do the wrong thing and will allow code that breaks when tools are upgraded.

aomarks replied:
Done. We now error if you don't import with  `.js` extension or the default module.

rictic commented on packages/transformers/src/lit-transformer.ts:
```diff
@@ -0,0 +1,263 @@
+          break;
+        }
+
+        default:
```
Recommend always wrapping cases in brackets, even when there's just one statement inside

aomarks replied:
Replaced it with `if` because there are only two cases and it's not really more readable as a `switch` imo.

rictic commented on packages/transformers/src/lit-transformer.ts:
```diff
@@ -0,0 +1,263 @@
+        continue;
+      }
+      const decoratorName = decorator.expression.expression.getText();
+      const visitors = this._classDecoratorVisitors.get(decoratorName) ?? [];
```
Here and elsewhere, I think we should look to see where this was imported from, and only apply our transform if we're confident it was imported from lit.

I think all you need to do for this is get the symbol for decorator.expression.expression, then get its first declaration. That should be a node inside the import, for decorators we want to transform.

aomarks replied:
Done. Also made it understand `import {property as potato}` aliases, and only remove bindings that we actually are transforming (when importing from `lit-element` we especially need this, since `LitElement` etc. are imported from the same module).

rictic commented on packages/transformers/src/lit-transformer.ts:
```diff
@@ -0,0 +1,263 @@
+      ) {
+        continue;
+      }
+      const decoratorName = decorator.expression.expression.getText();
```
Avoid calling getText(), as it's present on every node but isn't generally what you want (and it can throw for artificial nodes). We actually added a linter check to prevent calling getText in grimlock

We know that decorator.expression.expression is an Identifier so we can just get its text with the `text` property

aomarks replied:
Done, removed all usage.

rictic commented on packages/transformers/src/tests/idiomatic-test.ts:
```diff
@@ -0,0 +1,392 @@
+  class MyElement extends LitElement {
+    constructor() {
+      super(...arguments);
+      this._onClick = {
```
Hm, I don't think this is a safe change. Before _onClick was a function, now it's an object.

aomarks replied:
Hmm true, this works well for a Lit `@event` binding, but it wouldn't work if the function was invoked in any other way, which we can't too easily know.

Another option is:

```ts
class MyElement extends LitElement {
  _onClick(event) {
    console.log('click', event.target);
  }
}
Object.assign(MyElement.prototype._onClick, {once: true});
```

cc @justinfagnani since we discussed this a bit in the meeting

aomarks replied:
Hmm true, this works well for a Lit `@event` binding, but it wouldn't work if the function was invoked in any other way, which we can't too easily know.

Another option is:

```ts
class MyElement extends LitElement {
  _onClick(event) {
    console.log('click', event.target);
  }
}
Object.assign(MyElement.prototype._onClick, {once: true});
```

cc @justinfagnani since we discussed this a bit in the meeting

rictic replied:
Why not just:

```ts
MyElement.prototype._onClick.once = true;
```

aomarks replied:
Yep that works too -- though I could see an argument that it could be nice for our idiomatic style to have a slightly closer match between the way you set event handler options in e.g. `addEventListener(..., {once: true, capture: true})` using an object expression.

aomarks replied:
Switched to e.g. `@click=${{handleEvent: (e) => this._onClick(e), capture: true}}` as long as the handler is `private`, see comment from Kevin (and offline discussion). If not `private`, back to modifying the prototype method directly after the class declaration.

rictic commented on packages/transformers/src/tests/compile-ts-fragment.ts:
```diff
@@ -0,0 +1,153 @@
+ * Filesystem caching for the TypeScript CompilerHost implementation used in
+ * `compileTsFragment`. Shared across multiple invocations to significantly
+ * speed up the loading of common files (e.g. TypeScript core lib d.ts files).
+ * Assumes files on disk never change.
```
Consider calling this out a bit more, I almost missed it. Maybe put it on its own line and do:

```
**Important Note**: Assumes files on disk never change
```

Maybe name it something that indicates that its caching makes assumptions, like `OptimisticallyImmutableCompilerHostCache` or `OverlyOptimisticCompilerHostCache`

aomarks replied:
FYI just realized when I re-organized the commits I forgot to delete the other version of this file. So, this is not actually a new file, it just moved from `localize-tools` (though your comment is still valid). Added the delete to the newest commit since I don't want to invalidate the commits with comments anymore.

aomarks replied:
Added an "Important!" comment with explanation. Also just in case this wasn't clear, note this is in the test directory and used to speed up loading standard TS libs for unit tests.

kevinpschaaf commented:
Suggestion for more idiomatic TS -> JS without using decorators.
```
<div @click=${{passive: true, handleEvent: e => this.handleClick(e)}>
```

aomarks commented:
> Suggestion for more idiomatic TS -> JS without using decorators.
> 
> ```
> <div @click=${{passive: true, handleEvent: e => this.handleClick(e)}>
> ```

Done, as long as the method is `private`. Otherwise, we'll annotate the prototype method directly (in case it's e.g. used by a subclass).

aomarks commented:
@rictic @justinfagnani @kevinpschaaf PTAL

I've addressed all comments, and made some additional improvements. Main items:

- The `@eventOptions` decorator now transforms corresponding references inside Lit template event binding to the canonical syntax we agreed on offline (`@click=${{handleEvent: (e) => this._onClick(e), capture: true}}`), as long as the method is `private`. If it's not `private`, we annotate the prototype method directly (in case e.g. it is used in an event binding by a subclass).

- We now actually track imports from Lit modules, and check for decorators/html function with a type checker symbol lookup, instead of making naming assumptions.

- We now skip full traversal of files if we don't see any relevant Lit imports right away.

justinfagnani commented on packages/transformers/src/tests/idiomatic-test.ts:
```diff
@@ -0,0 +1,657 @@
+      this.str = "foo";
+      this.num = 42;
+    }
+    static get properties() {
```
we should put the properties block above the constructor.

justinfagnani replied:
Ideally there's some space around the class members. Is that possible?

justinfagnani commented on packages/transformers/src/tests/idiomatic-test.ts:
```diff
@@ -0,0 +1,657 @@
+      this.str = "foo";
+      this.num = 42;
+    }
+    static get properties() {
```
Can we have an option to email `static properties = {...}`?

aomarks replied:
Discussed offline. Decided to leave as the getter for now, since that's what we document for JS at https://lit.dev/. Can revisit later.

justinfagnani commented on packages/transformers/src/tests/idiomatic-test.ts:
```diff
@@ -0,0 +1,657 @@
+      super();
+      this.num = 42;
+    }
+    static get properties() {
```
We shouldn't move the properties block and/or constructor if they already exist.

justinfagnani commented on packages/transformers/src/tests/idiomatic-test.ts:
```diff
@@ -0,0 +1,657 @@
+    }
+    static get properties() {
+      return {
+        num: {state: true, attribute: false},
```
`state: true` implies `attribute: false` already.

justinfagnani replied:
Oh, that'd be good too.

aomarks replied:
Sent separately in https://github.com/lit/lit/pull/2030

justinfagnani commented on packages/transformers/src/tests/idiomatic-test.ts:
```diff
@@ -0,0 +1,657 @@
+  import {LitElement} from 'lit';
+  class MyElement extends LitElement {
+    get span() {
+      if (this.__span === undefined) {
+        this.__span = this.renderRoot?.querySelector('#mySpan');
+      }
+      return this.__span;
```
```suggestion
      return this.__span ??= this.renderRoot?.querySelector('#mySpan');
```

aomarks reviewed (commented):
All comments addressed @justinfagnani PTAL

justinfagnani commented on packages/transformers/src/lit-transformer.ts:
```diff
@@ -0,0 +1,398 @@
+const unreachable = (x: never) => x;
+
+/**
+ * Configurable transformer for Lit.
```
by "coalesce visitors" I mean that it's possible sometimes to write a batching visitor that calls visit on each of a set of visitors so they all go in one walk, rather than a series of walks. I think your code does this already.

justinfagnani commented on packages/transformers/src/visitor.ts:
```diff
@@ -0,0 +1,42 @@
+  | MemberDecoratorVisitor
+  | GenericVisitor;
+
+/** Visit a Lit decorator that is applied to a class. */
```
I meant as opposed to a decorated field, and for some reason when I read it before it seemed ambiguous to me and now doesn't. nvm!

justinfagnani commented on packages/transformers/src/lit-transformer.ts:
```diff
@@ -0,0 +1,398 @@
+      return false;
+    }
+    const specifier = node.moduleSpecifier.text;
+    if (
```
yeah. +1
